queryscript/src/compile/connection.rs

```diff
@@ -0,0 +1,211 @@
+use sqlparser::ast as sqlast;
+use std::path::{Path, PathBuf};
+use std::sync::Arc;
+use url::Url;
+
+use crate::ast::{self, Ident, Located, SourceLocation, ToSqlIdent};
+
+use super::compile::ExternalTypeRank;
+use super::external::schema_infer_expr_fn;
+use super::generics::{ExternalType, GenericConstructor};
+use super::inference::mkcref;
+use super::schema::{
+    CRef, Decl, Entry, Expr, MType, SQLBody, SQLNames, SQLSnippet, SType, STypedExpr,
+};
+use super::sql::select_limit_0;
+use super::{
+    error::Result,
+    schema::{DeclMap, ExprEntry},
+    CompileError,
+};
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct ConnectionString(Located<Url>);
+
+impl ConnectionString {
+    pub fn maybe_parse(
+        folder: Option<String>,
+        p: &str,
+        loc: &SourceLocation,
+    ) -> Result<Option<Arc<ConnectionString>>> {
+        let mut url = match Url::parse(p) {
+            Ok(url) => url,
+            Err(_) => return Ok(None),
+        };
+
+        match url.scheme() {
+            "file" => {
+                // NOTE: Eventually file URLs can allow you to reference schemas located
+                // elsewhere on the filesystem.
+                return Err(CompileError::unimplemented(
+                    loc.clone(),
+                    "file:// URLs are not supported. Use a relative path instead.",
+                ));
+            }
+            "duckdb" => {}
+            other => {
+                return Err(CompileError::unimplemented(
+                    loc.clone(),
+                    &format!("{} URLs", other),
+                ));
+            }
+        };
+
+        if matches!(url.scheme(), "duckdb") && matches!(url.host_str(), Some(_)) {
```

Comment:
Discussed offline: duckdb is a scheme instead of just having it recognize the file extension because a single file can contain multiple tables.  The same will be true of sqlite, etc.

If/when we support these kinds of files in locations other than the local filesystem, we'll find a different way of representing that (e.g. `connect('s3://foo.db', engine => 'duckdb')`).